core/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -338,7 +344,18 @@ public ActionModule(boolean ingestEnabled, boolean transportClient, Settings set
         autoCreateIndex = transportClient ? null : new AutoCreateIndex(settings, clusterSettings, resolver);
         destructiveOperations = new DestructiveOperations(settings, clusterSettings);
         Set<String> headers = actionPlugins.stream().flatMap(p -> p.getRestHeaders().stream()).collect(Collectors.toSet());
-        restController = new RestController(settings, headers);
+        UnaryOperator<RestHandler> restWrapper = null;
+        for (ActionPlugin plugin : actionPlugins) {
+            UnaryOperator<RestHandler> newRestWrapper = plugin.getRestHandlerWrapper(threadPool.getThreadContext());
+            if (newRestWrapper != null) {
+                logger.debug("Using REST wrapper from plugin " + plugin.getClass().getName());
+                if (restWrapper != null) {
```

Comment:
it's no secret that we use this for security, in such a case we really want to make sure nobody else has registered one. That is also true for anybody else that does stuff like this if you are the one protecting the handler with auth/autz you want to be sure nothing is executed before you. Now that said, I think as a compromise we could have 1 wrapper that is the primary wrapper and a set of wrappers that are unordered. Yet I still think we should try and find out if that is really needed and the only way is to wait until folks come to us with this. We can usually react pretty quickly on those things ie the compromise can come once it's a problem. I am not too concerned about this on that low level we are operation on.